Introduce annotation to skip subnets

michaellzc opened the pull request:
The goal is to support configuring the
`--subnets-to-ignore` flag in proxy-init

Add a new annotation `config.linkerd.io/skip-subnets` which
takes a comma-separated list of valid CIDR.
The argument will get map to the `--subnets-to-ignore`
flag in the proxy-init initContainer.

This can be validated with the provided integration test. Or
apply the following manifest
```yaml
kind: Pod
apiVersion: v1
metadata:
  name: debug-linkerd-dind
  annotations:
    linkerd.io/inject: enabled
    kubectl.kubernetes.io/default-container: docker
    config.linkerd.io/skip-subnets: "172.17.0.0/16"
spec:
  containers:
    - name: docker
      image: docker:19.03.5-dind
      imagePullPolicy: IfNotPresent
      securityContext:
        privileged: true
        runAsUser: 0
        capabilities:
            add:
            - NET_ADMIN
      resources: {}
```
Check the following
- the `--subnets-to-ignore` flag is added to `proxy-init`
- run `docker run  --rm curlimages/curl:7.78.0 -L -vv http://curl.haxx.se` inside the contaienr and it should work
- run `iptables -t nat --list --line-numbers` the iptables, there should be a rule in the `PROXY_INIT_REDIRECT` "return" on the provided subnet, learn more from https://github.com/linkerd/linkerd2-proxy-init/pull/53

Fixes #6758

Signed-off-by: Michael Lin <mlzc@hey.com>

mateiidavid commented on charts/partials/templates/_proxy-init.tpl:
```diff
@@ -24,6 +24,12 @@ args:
+- --subnets-to-ignore
+- {{ .Values.proxyInit.skipSubnets | quote }}
+{{- end }}
+{{- if .Values.proxyInit.skipOutboundPorts }}
+{{- end }}
```
Think this might've been added by accident?

mateiidavid reviewed (commented):
Thanks for the work @michaellzc, this looks good to me. I think we might have to clean up the `skipOutboundPorts` part in the template partial. 

It would also be really good if we could document this flag somewhere in code; perhaps in `values.yaml` and the corresponding helm doc (i.e the chart README). @alpeb might have an opinion here.

For the record, I used your manifest to test this (thanks for providing that):

```
  initContainers:
  - args:
    - --incoming-proxy-port
    - "4143"
    - --outgoing-proxy-port
    - "4140"
    - --proxy-uid
    - "2102"
    - --inbound-ports-to-ignore
    - 4190,4191,4567,4568
    - --outbound-ports-to-ignore
    - 4567,4568
    - --subnets-to-ignore
    - 172.17.0.0/16
```

the flag's picked up correctly. 

curl also seemed to work. Once we take care of those docs nits I'm good to merge this :D

michaellzc commented:
> Thanks for the work @michaellzc, this looks good to me. I think we might have to clean up the `skipOutboundPorts` part in the template partial.
> 
> It would also be really good if we could document this flag somewhere in code; perhaps in `values.yaml` and the corresponding helm doc (i.e the chart README). @alpeb might have an opinion here.
> 
> For the record, I used your manifest to test this (thanks for providing that):
> 
> ```
>   initContainers:
>   - args:
>     - --incoming-proxy-port
>     - "4143"
>     - --outgoing-proxy-port
>     - "4140"
>     - --proxy-uid
>     - "2102"
>     - --inbound-ports-to-ignore
>     - 4190,4191,4567,4568
>     - --outbound-ports-to-ignore
>     - 4567,4568
>     - --subnets-to-ignore
>     - 172.17.0.0/16
> ```
> 
> the flag's picked up correctly.
> 
> curl also seemed to work. Once we take care of those docs nits I'm good to merge this :D

I added the new flag to `values.yaml` and ran `bin/helm-docs` to update the README.md.

I also updated the `cli/cmd/doc.go` file. Assuming this is where the magic annotation data comes from 
- https://github.com/linkerd/website/blob/main/linkerd.io/content/2.11/reference/proxy-configuration.md?plain=1#L12
- https://github.com/linkerd/website/pull/996

mateiidavid reviewed (approved):
Awesome, thanks @michaellzc 🚢 📦

alpeb reviewed (approved):
Looks good, thanks @michaellzc :+1: 

One piece missing is having this new argument be included into the linkerd-cni plugin, but I'm not blocking this PR on it, because it's a different scenario... I'm not sure if that plugin is relevant for the DinD use case, but it'd be nice to include the new flag just for completeness sake. Would you be up to tackle that in a separate PR? It's just a matter of wiring the new setting in the `charts/linkerd2-cni` chart, and consider it in `cni-plugin/main.go` to be forwarded into the iptables library.

michaellzc commented:
@alpeb makes sense. 

Base on my limited knowledge of how CNI plug-in work, I think it still relies on setting up some up iptables rules. It's just being done by a global plug-in deployment instead of per pod inside the init container. Therefore. the dind issue is also applicable when CNI is being used.

I can make another pr for it.

mateiidavid commented:
Cool, thanks @michaellzc for taking on the work for the cni plugin too. Since it'll be a separate PR, I'm going to go ahead and merge this.
